File: python/src/com/sweetmanor/demo/GeneCompressed.py
# 《算法精粹》
# 基因压缩算法

class CompressedGene:

    def __init__(self, gene: str) -> None:
        self._compress(gene)

    def __str__(self) -> str:
        return self.decompress()

    def _compress(self, gene: str) -> None:
        # 数字不能以 0 开头，此处在开头添加一个 1
        self.bitString: int = 1

        for c in gene.upper():
            self.bitString <<= 2  # 左移2位
            if c == 'A':
                self.bitString |= 0b00
            elif c == 'C':
                self.bitString |= 0b01
            elif c == 'G':
                self.bitString |= 0b10
            elif c == 'T':
                self.bitString |= 0b11
            else:
                raise ValueError("Invalid Nucleotide:{}".format(c))

    def decompress(self) -> str:
        gene: str = ""

        # 逆序取值拼装，舍弃最后1位 1
        for i in range(0, self.bitString.bit_length() - 1, 2):
            bits: int = self.bitString >> i & 0b11  # 每次取2位
            if bits == 0b00:
                gene += 'A'
            elif bits == 0b01:
                gene += 'C'
            elif bits == 0b10:
                gene += 'G'
            elif bits == 0b11:
                gene += 'T'
            else:
                raise ValueError("Invalid bits:{}".format(bits))

        return gene[::-1]  # 反序


if __name__ == '__main__':
    from sys import getsizeof

    original: str = "TAGGGATTAACCGTTATATATATATAGCCATGGATCGATTATATAGGGATTAACCGTTATATATATATAGCCATGGATCGATTATA" * 100
    print("original is {} bytes".format(getsizeof(original)))

    # compress
    compressed: CompressedGene = CompressedGene(original)
    print("compressed is {} bytes".format(getsizeof(compressed.bitString)))
    print(compressed)

    # decompress
    print("original and decompressed are the same: {}".format(original == compressed.decompress()))

File: java/src/main/java/com/sweetmanor/demo/GeneCompressed.java
package com.sweetmanor.demo;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.BitSet;

/**
 * 基因序列压缩 —— 《算法精粹》
 *
 * @author ijlhjj
 * @version 1.0 2023-12-09
 */
public class GeneCompressed {
    private static final Logger logger = LogManager.getLogger();

    private BitSet bitSet;
    private int length; //存储字符串长度

    public GeneCompressed(String gene) {
        compress(gene);
    }

    /**
     * 压缩方法：将字符串转换后存储到 bitSet
     */
    private void compress(String gene) {
        length = gene.length(); //字符串长度
        bitSet = new BitSet(length * 2);//2位存储一个字符

        String upperGene = gene.toUpperCase();
        for (int i = 0; i < length; i++) {
            int first = 2 * i;
            int second = first + 1;
            switch (upperGene.charAt(i)) {
                case 'A': //00
                    bitSet.set(first, false);
                    bitSet.set(second, false);
                    break;
                case 'C': //01
                    bitSet.set(first, false);
                    bitSet.set(second, true);
                    break;
                case 'G': //10
                    bitSet.set(first, true);
                    bitSet.set(second, false);
                    break;
                case 'T': //11
                    bitSet.set(first, true);
                    bitSet.set(second, true);
                    break;
                default:
                    throw new IllegalArgumentException("The provided gene String contains characters other than ACGT");
            }
        }
    }

    /**
     * 解压缩方法
     */
    public String decompress() {
        if (bitSet == null)
            return "";

        StringBuilder sb = new StringBuilder(length);
        for (int i = 0; i < length * 2; i += 2) {
            int firstBit = bitSet.get(i) ? 1 : 0;
            int secondBit = bitSet.get(i + 1) ? 1 : 0;
            int bits = firstBit << 1 | secondBit; //组合两个数位

            switch (bits) {
                case 0b00: // 00 - A
                    sb.append('A');
                    break;
                case 0b01: // 01 - C
                    sb.append('C');
                    break;
                case 0b10: // 10 - G
                    sb.append('G');
                    break;
                case 0b11: // 11 - T
                    sb.append('T');
                    break;
                default:
                    break;
            }
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        final String original = "TAGGGATTAACCGTTATATATATATAGCCATGGATCGATTATATAGGGATTAACCGTTATATATATATAGCCATGGATCGATTATA";
        GeneCompressed compressed = new GeneCompressed(original);
        final String decompressed = compressed.decompress();
        logger.info(decompressed);
        logger.info("original is the same as decompressed:{} ", original.equalsIgnoreCase(decompressed));
    }

}
